src/lib.rs

```diff
@@ -157,35 +157,35 @@ impl<T: CoordinateType> PiecewiseLinearFunction<T> {
             Err(idx) => {
                 if idx == 0 || idx == self.coordinates.len() {
                     // Outside the function's domain
-                    return None;
+                    return Err(Error::OutsideDomain);
                 } else {
                     idx
                 }
             }
         };
 
         if idx == 0 {
-            Some(Line::new(self.coordinates[idx], self.coordinates[idx + 1]))
+            Ok(Line::new(self.coordinates[idx], self.coordinates[idx + 1]))
         } else {
-            Some(Line::new(self.coordinates[idx - 1], self.coordinates[idx]))
+            Ok(Line::new(self.coordinates[idx - 1], self.coordinates[idx]))
         }
     }
 
     /// Computes the value f(x) for this piecewise linear function.
     ///
     /// Returns `None` if `x` is outside the domain of f.
-    pub fn y_at_x(&self, x: T) -> Option<T> {
+    pub fn y_at_x(&self, x: T) -> Result<T, Error> {
```

Comment:
Why aren't we also returning return Err(Error::OutsideDomain); if 'x' is outside the domain in this method?

Reply:
It's passed up from the call to `segment_at_x`, which is the underlying reason why it may error in the first place

Reply:
ah right, thanks.